pkg/services/ngalert/state/cache.go

```diff
@@ -330,6 +330,36 @@ func (c *cache) removeByRuleUID(orgID int64, uid string) []*State {
 	return states
 }
 
+// asInstances returns the whole content of the cache as a slice of AlertInstance.
+func (c *cache) asInstances(skipNormalState bool) []ngModels.AlertInstance {
+	var states []ngModels.AlertInstance
+	c.mtxStates.RLock()
+	defer c.mtxStates.RUnlock()
+	for _, orgStates := range c.states {
+		for _, v1 := range orgStates {
+			for _, v2 := range v1.states {
+				if skipNormalState && IsNormalStateWithNoReason(v2) {
+					continue
+				}
+				key, err := v2.GetAlertInstanceKey()
+				if err != nil {
+					continue
+				}
+				states = append(states, ngModels.AlertInstance{
+					AlertInstanceKey:  key,
+					Labels:            ngModels.InstanceLabels(v2.Labels),
+					CurrentState:      ngModels.InstanceStateType(v2.State.String()),
+					CurrentReason:     v2.StateReason,
+					LastEvalTime:      v2.LastEvaluationTime,
+					CurrentStateSince: v2.StartsAt,
+					CurrentStateEnd:   v2.EndsAt,
```

Comment:
I noticed `ResultFingerprint` isn't copied, do we need to?

Reply:
Good catch!